pkg/frontend/querymiddleware/querysharding.go

```diff
@@ -246,47 +227,104 @@ func mapEngineError(err error) error {
 	return apierror.New(errorType, cause.Error())
 }
 
-// shardQuery attempts to rewrite the input query in a shardable way. Returns the rewritten query
-// to be executed by PromQL engine with shardedQueryable or an empty string if the input query
-// can't be sharded.
-func (s *querySharding) shardQuery(ctx context.Context, query string, totalShards int) (string, *astmapper.MapperStats, error) {
-	stats := astmapper.NewMapperStats()
-	ctx, cancel := context.WithTimeout(ctx, shardingTimeout)
-	defer cancel()
+type querySharder struct {
+	squasher          astmapper.Squasher
+	limit             Limits
+	maxSeriesPerShard uint64
+	logger            log.Logger
+	metrics           queryShardingMetrics
+}
 
-	squasher := astmapper.EmbeddedQueriesSquasher
-	summer, err := astmapper.NewQueryShardSummer(totalShards, squasher, s.logger, stats)
+func newQuerySharder(
+	limit Limits,
+	maxSeriesPerShard uint64,
+	reg prometheus.Registerer,
+	logger log.Logger,
+) *querySharder {
+	return &querySharder{
+		squasher:          astmapper.EmbeddedQueriesSquasher,
+		limit:             limit,
+		maxSeriesPerShard: maxSeriesPerShard,
+		metrics:           newQueryShardingMetrics(reg),
+		logger:            logger,
+	}
+}
+
+// shard attempts to rewrite expr in a sharded way.
+//
+// expr may be modified in place, including if an error is returned.
+//
+// If the query can't be sharded, shard returns nil and no error, and expr is unchanged.
+func (s *querySharder) shard(ctx context.Context, expr parser.Expr, requestedShardCount int, seriesCount *EstimatedSeriesCount, totalQueries int32) (parser.Expr, error) {
+	log := spanlogger.FromContext(ctx, s.logger)
+	log.DebugLog("msg", "attempting query sharding", "query", expr)
+
+	tenantIDs, err := tenant.TenantIDs(ctx)
```

Comment:
I would expect the middleware to extract tenant IDs from the context and pass them explicitly as an argument to this method.